compiler/sem/semexprs.nim

```diff
@@ -319,6 +319,8 @@ proc semConv(c: PContext, n: PNode): PNode =
   var op = semExprWithType(c, n[1])
   if targetType.kind != tyGenericParam and targetType.isMetaType:
     let final = inferWithMetatype(c, targetType, op, true)
+    # XXX: this makes little sense -- the source and target type of the
+    #      resulting conversion are the same
```

Comment:
My guess is that this _might_ be when going from a given type to `untyped` or `typed`. See this test case: https://github.com/nim-works/nimskull/blob/devel/tests/lang_stmts/bind/tinvalidbindtypedesc.nim

I haven't traced it in detail, but `echo`'s params are `varargs[typed, $]`, that would explain why they had to make it act like an identity here to get the code to "work".

Reply:
It's something arising from `inferWithMetatype` always overwriting the result node's type. Consider the follow (which takes the `inferWithMetatype` path):
```nim
type ArrayTy[T] = distinct array[1, T]

discard ArrayTy(["str"])
```

`inferWithMetatype` sets the type of the expression `["str"]` to `ArrayTy[string]`, but here in `semConv`, it is then wrapped with `nkConv` node also using `ArrayTy[string]` as the type.

The more correct thing to do would be leaving the type of the processed expression in `inferWithMetatype` as is (at least if `coerceDistincts == true`), and only use the inferred type for the `nkConv` expression.

Reply:
I strongly agree, and would perhaps go a bit further; I think this PR is all good so going to bors.

---

### Aside: Going a Bit Further

The key points for me, and I believe entirely line up with yours are:
1. the type of expression `["str"]` should always be the same, something like: `array[1, static[string]]`
2. applying `ArrayTy` is a few steps, but the key conversion expression result's type is `ArrayTy[static[string]]`

Coercion of distincts should determine whether the conversion expression result's type is an error (because distinct coercion is disallowed) or successful and therefore `ArrayTy[static[string]]`.

Definitely agree that `inferWithMetatype` is applying the wrong type on `["str"]`, I think this is a generally unhygienic habit of not creating intermediate nodes and apply appropriate types and conversions all over the compiler.

Reply:
Yep, they entirely line up.